mindmeld/models/nn_utils/input_encoders.py

```diff
@@ -0,0 +1,937 @@
+        msg = f"Subclass {self.name} need to implement this method"
+        raise NotImplementedError(msg)
+
+    def _load_bert_embedder_from_checkpoint(self, pretrained_model_name_or_path):
+        model_bunch = HuggingfaceTransformersContainer(
+            pretrained_model_name_or_path, reload=True).get_model_bunch()
+        self._bert_config = model_bunch.config
+        self._bert_tokenizer = model_bunch.tokenizer
+        self._bert_model = model_bunch.model.to(self.device)
+        if not self.update_embeddings:
+            for param in self._bert_model:
+                param.requires_grad = False
+
+    def _dump_bert_embedder_to_checkpoint(self, ckpt_folder):
+        os.makedirs(ckpt_folder, exist_ok=True)
+        self._bert_config.save_pretrained(ckpt_folder)
+        self._bert_tokenizer.save_pretrained(ckpt_folder)
+        self._bert_model.save_pretrained(ckpt_folder)
+
+
+# sub-classes that define a custom `batch_encode` on top of base encoders
+
+
+class SeqClsEncoderForEmbLayer(WordEmbeddingsBasedEncoder):
+    """This class produces encoding for the given textual input as a sequence of ids.
+    The inputs can be singular or batched.
+    """
+
+    def batch_encode(
+        self,
+        examples: Union[str, List[str]],
+        labels: Union[int, List[int]] = None,
+        padding_length: int = None,
+        add_terminals: bool = False
+    ):
+        """
+        Returns batched encodings that can be used as an input to a featurizer
+        """
+
+        examples, labels = self._reformat_and_validate(examples, labels)
+
+        # convert to tokens and obtain sequence lengths
+        examples = [self._tokenizer(text) for text in examples]
+        sequence_lengths = [len(seq) for seq in examples]
+
+        # if padding_length is None, it is computed as max(length of all seqs from inputted text)
+        padding_length = padding_length or self.padding_length
+        # add 2 for start and end tokens respectively if add_terminals is True
+        add_terminals = add_terminals or self.add_terminals
+        curr_max = max(sequence_lengths) + 2 if add_terminals else max(sequence_lengths)
+        padding_length = min(padding_length, curr_max) if padding_length else curr_max
+
+        # batchify by truncating or padding for each example
+        seq_ids, sequence_lengths = zip(*[
+            self.encode(example, padding_length, add_terminals) for example in examples
+        ])
+
+        # create output dict
+        return_dict = {
+            "seq_ids": torch.as_tensor(seq_ids, dtype=torch.long),
+            "seq_lengths": torch.as_tensor(sequence_lengths, dtype=torch.long)
+        }
+
+        if labels:
+            return_dict.update({
+                "labels": torch.as_tensor(labels, dtype=torch.long)
+            })
+
+        return return_dict
+
+
+class TokenClsEncoderForEmbLayer(WordEmbeddingsBasedEncoder):
+
+    def _reformat_and_validate(self, examples, labels=None):
```

Comment:
Can't we move this to `PyTorchTaggerModels` since this only changes with the task and not the Encoder type?